File: src/main/java/io/github/starwishsama/namelessbot/objects/dtos/RssItem.java
package io.github.starwishsama.namelessbot.objects.dtos;

import com.rometools.rome.feed.synd.SyndEntry;
import com.rometools.rome.feed.synd.SyndFeed;
import com.rometools.rome.io.FeedException;
import com.rometools.rome.io.SyndFeedInput;
import com.rometools.rome.io.XmlReader;
import lombok.Data;

import java.io.IOException;
import java.net.URL;
import java.util.List;

@Data
public class RssItem {
    private String address;
    private boolean ifEnabled;
    private List<Long> subscribers;

    public RssItem(String address){
        this.address = address;
    }

    public static void main(String[] args){
    }

    public String getContext() {
        return (simplifyHTML(getFromURL(address)));
    }

    public String getTitle() {
        return (simplifyHTML(getTitleFromURL(address)));
    }

    public SyndEntry getEntry(){
        return getEntryFromURL(address);
    }

    // 此函数仅供内部调用，正常情况下不应调用
    private static String getFromURL(String address) {
        try {
            URL url = new URL(address);
            // 读取RSS源
            XmlReader reader = new XmlReader(url);
            SyndFeedInput input = new SyndFeedInput();
            // 得到SyndFeed对象，即得到RSS源里的所有信息
            SyndFeed feed = input.build(reader);
            // 得到Rss新闻中子项列表
            List<SyndEntry> entries = feed.getEntries();
            SyndEntry entry = entries.get(0);
            return (entry.getTitle() + "\n" + "-------------------------\n" + entry.getDescription().getValue().trim());
        } catch (Exception e) {
            e.printStackTrace();
            return "Encountered a wrong URL or a network error.";
        }
    }

    private static SyndEntry getEntryFromURL(String address) {
        try {
            URL url = new URL(address);
            // 读取RSS源
            XmlReader reader = new XmlReader(url);
            SyndFeedInput input = new SyndFeedInput();
            // 得到SyndFeed对象，即得到RSS源里的所有信息
            SyndFeed feed = input.build(reader);
            // 得到Rss新闻中子项
            return feed.getEntries().get(0);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    private static String getTitleFromURL(String address){
        try {
            URL url = new URL(address);
            XmlReader reader = new XmlReader(url);
            SyndFeedInput input = new SyndFeedInput();
            // 得到SyndFeed对象，即得到RSS源里的所有信息
            SyndFeed feed = input.build(reader);
            // 得到Rss新闻中子项列表
            List<SyndEntry> entries = feed.getEntries();
            SyndEntry entry = entries.get(0);
            return entry.getTitle();
        } catch (Exception e){
            e.printStackTrace();
            return ("Encountered a wrong URL or a network error.");
        }
    }

    public static String simplifyHTML(String context) {
        context = context.replaceAll("<br />", "\n").replaceAll("<br>", "\n").replaceAll("</p><p>", "\n")
                .replaceAll("	", "");
        while (context.indexOf('<') != -1) {
            int l = context.indexOf('<');
            int r = context.indexOf('>');
            context = context.substring(0, l) + context.substring(r + 1);
        }
        while (context.contains("\n\n")) {
            context = context.replaceAll("\n\n", "\n");
        }
        return context;
    }

    public static SyndEntry getRSSItem(String address){
        try {
            URL url = new URL(address);
            XmlReader reader = new XmlReader(url);
            SyndFeedInput input = new SyndFeedInput();
            // 得到SyndFeed对象，即得到RSS源里的所有信息
            SyndFeed feed = input.build(reader);
            // 得到RSS源中子项列表
            if (feed != null)
                return feed.getEntries().get(0);
        } catch (FeedException | IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    public static List<SyndEntry> getRSSItems(String address){
        try {
            URL url = new URL(address);
            XmlReader reader = new XmlReader(url);
            SyndFeedInput input = new SyndFeedInput();
            // 得到SyndFeed对象，即得到RSS源里的所有信息
            SyndFeed feed = input.build(reader);
            // 得到RSS源中子项列表
            return feed.getEntries();
        } catch (FeedException | IOException e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: src/main/kotlin/io/github/starwishsama/nbot/objects/RssItem.kt
package io.github.starwishsama.nbot.objects

import com.rometools.rome.feed.synd.SyndEntry
import com.rometools.rome.feed.synd.SyndFeed
import com.rometools.rome.io.FeedException
import com.rometools.rome.io.SyndFeedInput
import com.rometools.rome.io.XmlReader
import java.io.IOException
import java.net.URL


class RssItem(private val address: String) {
    val ifEnabled = false
    val subscribers: List<Long>? = null
    val context: String
        get() = simplifyHTML(getFromURL(address))

    val title: String
        get() = simplifyHTML(getTitleFromURL(address))

    val entry: SyndEntry?
        get() = getEntryFromURL(address)

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
        }

        // 此函数仅供内部调用，正常情况下不应调用
        private fun getFromURL(address: String): String {
            return try {
                val url = URL(address)
                // 读取RSS源
                val reader = XmlReader(url)
                val input = SyndFeedInput()
                // 得到SyndFeed对象，即得到RSS源里的所有信息
                val feed: SyndFeed = input.build(reader)
                // 得到Rss新闻中子项列表
                val entries = feed.entries
                val entry = entries[0]
                entry.title + "\n" + "-------------------------\n" + entry.description.value.trim { it <= ' ' }
            } catch (e: Exception) {
                e.printStackTrace()
                "Encountered a wrong URL or a network error."
            }
        }

        private fun getEntryFromURL(address: String): SyndEntry? {
            try {
                val url = URL(address)
                // 读取RSS源
                val reader = XmlReader(url)
                val input = SyndFeedInput()
                // 得到SyndFeed对象，即得到RSS源里的所有信息
                val feed: SyndFeed = input.build(reader)
                // 得到Rss新闻中子项
                return feed.entries[0]
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return null
        }

        private fun getTitleFromURL(address: String): String {
            return try {
                val url = URL(address)
                val reader = XmlReader(url)
                val input = SyndFeedInput()
                // 得到SyndFeed对象，即得到RSS源里的所有信息
                val feed: SyndFeed = input.build(reader)
                // 得到Rss新闻中子项列表
                val entries = feed.entries
                val entry = entries[0]
                entry.title
            } catch (e: Exception) {
                e.printStackTrace()
                "Encountered a wrong URL or a network error."
            }
        }

        fun simplifyHTML(rssContext: String): String {
            var context = rssContext
            context = context.replace("<br />".toRegex(), "\n").replace("<br>".toRegex(), "\n")
                .replace("</p><p>".toRegex(), "\n")
                .replace("	".toRegex(), "")
            while (context.indexOf('<') != -1) {
                val l = context.indexOf('<')
                val r = context.indexOf('>')
                context = context.substring(0, l) + context.substring(r + 1)
            }
            while (context.contains("\n\n")) {
                context = context.replace("\n\n".toRegex(), "\n")
            }
            return context
        }

        fun getRSSItem(address: String?): SyndEntry? {
            try {
                val url = URL(address)
                val reader = XmlReader(url)
                val input = SyndFeedInput()
                // 得到SyndFeed对象，即得到RSS源里的所有信息
                val feed: SyndFeed = input.build(reader)
                // 得到RSS源中子项列表
                return feed.entries[0]
            } catch (e: FeedException) {
                e.printStackTrace()
            } catch (e: IOException) {
                e.printStackTrace()
            }
            return null
        }

        fun getRSSItems(address: String?): List<SyndEntry>? {
            return try {
                val url = URL(address)
                val reader = XmlReader(url)
                val input = SyndFeedInput()
                // 得到SyndFeed对象，即得到RSS源里的所有信息
                val feed: SyndFeed = input.build(reader)
                // 得到RSS源中子项列表
                feed.entries
            } catch (e: FeedException) {
                e.printStackTrace()
                null
            } catch (e: IOException) {
                e.printStackTrace()
                null
            }
        }
    }

}